src/Microsoft.DotNet.ProjectModel/Workspace/WorkspaceContext.cs

```diff
@@ -0,0 +1,328 @@
+            if (project == null)
+            {
+                // project doesn't exist anymore
+                currentEntry.Reset();
+                return currentEntry;
+            }
+
+            if (currentEntry.HasChanged)
+            {
+                var builder = new ProjectContextBuilder()
+                    .WithProjectResolver(path => GetProject(path))
+                    .WithLockFileResolver(path => GetLockFile(path))
+                    .WithProject(project)
+                    .WithTargetFramework(framework);
+
+                currentEntry.ProjectContext = builder.Build();
+            }
+
+            return currentEntry;
+        }
+
+        private ProjectContextEntry AddProjectContextEntryWithDependency(string projectDirectory,
+                                                                         NuGetFramework framework,
+                                                                         ProjectContextEntry currentEntry)
+        {
+            var entry = AddProjectContextEntry(projectDirectory, framework, currentEntry);
+            if (!entry.HasDependencyResolved)
+            {
+                entry.ResolveDependencies(Configuration);
+            }
+
+            return entry;
+        }
+
+        private class FileModelEntry<TModel> where TModel : class
+        {
+            private DateTime _lastWriteTime;
+
+            public TModel Model { get; set; }
+
+            public string FilePath { get; set; }
+
+            public void UpdateLastWriteTime()
+            {
+                _lastWriteTime = File.GetLastWriteTime(FilePath);
+            }
+
+            public bool IsInvalid
+            {
+                get
+                {
+                    if (Model == null)
+                    {
+                        return true;
+                    }
+
+                    if (!File.Exists(FilePath))
+                    {
+                        return true;
+                    }
+
+                    return _lastWriteTime < File.GetLastWriteTime(FilePath);
+                }
+            }
+
+            public void Reset()
+            {
+                Model = null;
+                FilePath = null;
+                _lastWriteTime = DateTime.MinValue;
+            }
+        }
+
+        private static List<string> PathResolve(string projectPath)
+        {
+            if (File.Exists(projectPath))
+            {
+                var filename = Path.GetFileName(projectPath);
+                if (!Project.FileName.Equals(filename, StringComparison.OrdinalIgnoreCase) &&
+                    !GlobalSettings.FileName.Equals(filename, StringComparison.OrdinalIgnoreCase))
+                {
+                    return null;
+                }
+
+                projectPath = Path.GetDirectoryName(projectPath);
+            }
+
+            if (File.Exists(Path.Combine(projectPath, Project.FileName)))
+            {
+                return new List<string> { projectPath };
```

Comment:
This should probably use `ProjectRootResolver` as well and find the `global.json` for the project. If I run `WorkspaceContext.CreateFrom('/path/to/project.json`)` and there's a `global.json` for that project, I'd want `WorkspaceContext` to give me the full context.

That means we can probably just remove this `if` and put the condition in as an OR for the one below.